R/particle_filter_state_nested.R

```diff
@@ -83,31 +84,55 @@ particle_filter_state_nested <- R6::R6Class(
       }
 
       if (!is.null(initial)) {
-        initial_data <- initial(model$info(), n_particles, pars)
-        if (is.list(initial_data)) {
-          steps <- particle_steps(steps, initial_data$step)
-          model$set_state(initial_data$state, initial_data$step)
+        initial_data <- Map(initial,
+          info = model$info(),
+          pars = pars,
+          MoreArgs = list(n_particles = n_particles)
+        )
+        if (!is.null(unlist(initial_data))) {
+          if (any(c(c("step", "state") %in% names(initial_data[[1]])))) {
+            init_step <- unlist(lapply(initial_data, "[[", "step"))
+            if (length(init_step) != n_particles * length(pars)) {
+              if (length(init_step) == length(pars)) {
+                init_step <- rep(init_step, each = n_particles)
+              } else {
+                stop(sprintf("Expected 'step' to be length %d or %d",
+                             length(pars), n_particles * length(pars)))
+              }
+            }
+            init_state <- unlist(lapply(initial_data, "[[", "state"))
+            if (!is.null(init_state)) {
+              init_state <- matrix(init_state, ncol = 2)
+            }
+            steps <- particle_steps(steps, init_step)
+            model$set_state(init_state, init_step)
+          } else {
+            model$set_state(list_to_array(initial_data))
+          }
         } else {
-          model$set_state(initial_data)
+          model$set_state(NULL)
         }
       }
 
-      index_data <- if (is.null(index)) NULL else index(model$info())
-      if (!is.null(compare) && !is.null(index_data$run)) {
-        model$set_index(index_data$run)
+      index_data <- if (is.null(index)) NULL else lapply(model$info(), index)
```

Comment:
We probably should check that all index entries are the same

Reply:
As with the `initial` comment above. Not sure if we currently allow a case where index entries could be different between populations

Reply:
We should not allow it, but enforce that they're identical unless that is too costly